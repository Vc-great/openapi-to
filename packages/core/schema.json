{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "@openapi-to/define-config",
  "title": "JSON schema for @openapi-to/core define-config",
  "type": "object",
  "required": [],
  "properties": {
    "root": {
      "type": "string",
      "description": "Project root directory. Can be an absolute path, or a path relative from the location of the config file itself.",
      "default": "process.cwd()"
    },
    "input": {
      "type": "object",
      "description": "",
      "properties": {
        "path": {
          "type": "string",
          "description": "Path to be used as the input. Can be an absolute path, or a path relative from the defined root option."
        }
      }
    },
    "output": {
      "type": "object",
      "description": "",
      "properties": {
        "path": {
          "type": "string",
          "description": "Path to be used to export all generated files. Can be an absolute path, or a path relative based of the defined root option."
        },
        "clean": {
          "type": "boolean",
          "description": "Remove previous generated files and folders."
        },
        "write": {
          "type": "boolean",
          "description": "Enabled or disable the writing to the filesystem. This is being used for the playground.",
          "default": "true"
        }
      }
    },
    "hooks": {
      "type": "object",
      "description": "Hooks that can be called when a specific action is done in openapi-to. ",
      "properties": {
        "done": {
          "oneOf": [
            {
              "type": "string",
              "description": "Hook that will be called at the end of all executions. Useful for running Prettier or ESLint to use your own linting structure."
            },
            {
              "type": "array",
              "items": {
                "type": "string"
              },
              "description": "Hook that will be called at the end of all executions. Useful for running Prettier or ESLint to use your own linting structure."
            }
          ]
        }
      }
    },
    "plugins": {
      "oneOf": [
        {
          "type": "array",
          "items": {
            "type": "array",
            "prefixItems": [
              {
                "type": "string"
              }
            ],
            "items": {
              "type": "object"
            }
          }
        },
        {
          "type": "object"
        }
      ],
      "description": "Plugins"
    }
  }
}
