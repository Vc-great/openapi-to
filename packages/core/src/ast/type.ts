import type { JSDocTagStructure } from "ts-morph";
import type {
  AmbientableNodeStructure,
  ExportableNodeStructure,
  JSDocableNodeStructure,
  Structure,
} from "ts-morph";
import type {
  BindingNamedNodeStructure,
  ExclamationTokenableNodeStructure,
  KindedStructure,
  OptionalKind,
  StructureKind,
  TypedNodeStructure,
  VariableDeclarationKind,
  WriterFunction,
} from "ts-morph";

export interface VariableStatementStructureOfObject
  extends Structure,
    VariableStatementSpecificStructure,
    JSDocableNodeStructure,
    AmbientableNodeStructure,
    ExportableNodeStructure {}

interface VariableStatementSpecificStructure
  extends KindedStructure<StructureKind.VariableStatement> {
  declarationKind?: VariableDeclarationKind;
  declarations: OptionalKind<VariableDeclarationStructure>[];
}

export interface VariableDeclarationStructure
  extends Structure,
    VariableDeclarationSpecificStructure,
    BindingNamedNodeStructure,
    InitializerExpressionableNodeStructure,
    TypedNodeStructure,
    ExclamationTokenableNodeStructure {}

interface VariableDeclarationSpecificStructure
  extends KindedStructure<StructureKind.VariableDeclaration> {}

export interface InitializerExpressionableNodeStructure {
  initializer?: string | WriterFunction | ObjectStructure[];
}

export interface ObjectStructure {
  key: string;
  value: string;
  docs: JSDocStructure[];
}

export interface JSDocStructure {
  /**
   * The description of the JS doc.
   * @remarks To force this to be multi-line, add a newline to the front of the string.
   */
  description?: string;
  /** JS doc tags (ex. `&#64;param value - Some description.`). */
  tags?: OptionalKind<JSDocTagStructure>[];
}

export type JSDoc = JSDocStructure[];
