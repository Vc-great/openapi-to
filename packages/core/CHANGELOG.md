# @openapi-to/core

## 3.0.1

### Patch Changes

- fix bug

## 3.0.0

### Major Changes

- fix bug
- fix: bug
- update
- refactor

## 3.0.0-alpha.5

### Major Changes

- fix: bug

## 3.0.0-alpha.4

### Major Changes

- fix bug

## 3.0.0-alpha.3

### Major Changes

- refactor

## 3.0.0-alpha.2

### Major Changes

- fix

## 2.4.0-alpha.1

### Patch Changes

- fix bug

## 2.4.0-alpha.0

### Minor Changes

- update all

## 2.3.0

### Minor Changes

- generate swr

## 2.2.0

### Minor Changes

- fix bug

## 2.1.0

### Minor Changes

- optimize request type zod

## 2.0.0

### Major Changes

- [`e519466`](https://github.com/Vc-great/openapi-to/commit/e5194667c7416e817a498d592c357a7ae9c05f22) Thanks [@Vc-great](https://github.com/Vc-great)! - v2

- [`a39a900`](https://github.com/Vc-great/openapi-to/commit/a39a9002dda434d8a65768f55c69875ed8ad1eea) - fix bug

- 2.0

- [`367be25`](https://github.com/Vc-great/openapi-to/commit/367be252aa434487c09c4566e77792839867b509) - fix bug

- [`7fd9e58`](https://github.com/Vc-great/openapi-to/commit/7fd9e58417ef5563dedf945fbb030b70b8b09bd6) - rc

## 2.0.0-rc.5

### Major Changes

- rc

## 2.0.0-alpha.4

### Major Changes

- fix bug

## 2.0.0-alpha.3

### Major Changes

- fix bug

## 2.0.0-alpha.2

### Major Changes

- v2
