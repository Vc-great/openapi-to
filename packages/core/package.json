{
  "name": "@openapi-to/core",
  "version": "2.4.0-alpha.1",
  "description": "Generator core",
  "keywords": [
    "typescript",
    "plugins",
    "codegen"
  ],
  "repository": {
    "type": "git",
    "url": "",
    "directory": "packages/core"
  },
  "license": "MIT",
  "author": "",
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs",
      "default": "./dist/index.cjs"
    },
    "./utils": {
      "import": "./dist/utils.js",
      "require": "./dist/utils.cjs",
      "default": "./dist/utils.cjs"
    },
    "./package.json": "./package.json",
    "./*": "./*"
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "src",
    "dist",
    "*.d.ts",
    "*.d.cts",
    "schema.json",
    "!/**/**.test.**",
    "!/**/__tests__/**"
  ],
  "scripts": {
    "build": "tsup",
    "clean": "npx rimraf ./dist",
    "lint": "bun biome lint .",
    "lint:fix": "bun biome lint --write --unsafe .",
    "release": "pnpm publish --no-git-check",
    "release:canary": "bash ../../.github/canary.sh && node ../../scripts/build.js canary && pnpm publish --no-git-check -tag canary",
    "release:alpha": "bash ../../.github/canary.sh && node ../../scripts/build.js alpha && pnpm publish --no-git-check -tag alpha",
    "start": "tsup --watch",
    "test": "vitest --passWithNoTests",
    "typecheck": "tsc -p ./tsconfig.json --noEmit --emitDeclarationOnly false"
  },
  "dependencies": {
    "@types/json-schema": "^7.0.15",
    "axios": "^1.6.7",
    "code-block-writer": "^12.0.0",
    "consola": "^3.2.3",
    "directory-tree": "^3.5.1",
    "do-swagger2openapi": "^7.0.18",
    "find-up": "^7.0.0",
    "fs-extra": "^11.2.0",
    "is-chinese": "^2.0.0",
    "is-url": "^1.2.4",
    "js-runtime": "^0.0.7",
    "json-schema": "^0.4.0",
    "lodash": "^4.17.21",
    "natural-orderby": "^3.0.2",
    "oas": "^24.9.0",
    "picocolors": "^1.0.0",
    "pinyin-pro": "^3.25.0",
    "seedrandom": "^3.0.5",
    "semver": "^7.5.4",
    "tinyrainbow": "^1.2.0",
    "ts-morph": "^23.0.0"
  },
  "devDependencies": {
    "@openapi-to/config-ts": "workspace:*",
    "@openapi-to/config-tsup": "workspace:*",
    "@types/fs-extra": "^11.0.4",
    "@types/is-url": "^1.2.32",
    "@types/lodash": "^4.17.9",
    "@types/seedrandom": "^3.0.8",
    "@types/semver": "^7.5.6",
    "ora": "^7.0.1"
  },
  "engines": {
    "node": ">=18",
    "pnpm": ">=8.3.0"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  }
}
