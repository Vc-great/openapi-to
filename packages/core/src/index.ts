export * from "./defineConfig";
export * from "./errors.ts";
export { Generator } from "./Generator.ts";
export { PackageManager } from "./PackageManager.ts";
// dprint-ignore
export { SchemaGenerator } from "./SchemaGenerator.ts";
export * from "./types.ts";
/*
export default build;*/
export { AST } from "./ast/ast.ts";
export { build, requestRemoteData } from "./build.ts";
export { OpenAPI } from "./openapi/OpenAPI.ts";
export { createPlugin } from "./plugins.ts";

/*



export interface _Register {}
export type Plugins = _Register;
export type OptionsPlugins = { [K in keyof Plugins]: Plugins[K]["options"] };
* */
