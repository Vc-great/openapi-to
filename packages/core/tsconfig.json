{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "types": ["vitest/globals", "bun-types"]
  },
  "include": ["src/**/*.ts", "./package.json", "vitest.config.ts", "tsup.config.ts", "../config-ts/reset.d.ts"]
}
