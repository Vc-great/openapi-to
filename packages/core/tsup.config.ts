import { optionsCJS, optionsESM } from "@openapi-to/config-tsup";

import { defineConfig } from "tsup";

export default defineConfig([
  {
    ...optionsCJS,
    noExternal: [/find-up/],
  },
  optionsESM,
  {
    ...optionsCJS,
    entry: {
      utils: "src/utils/index.ts",
    },
    name: "utils",
    noExternal: [/find-up/],
  },
  {
    ...optionsESM,
    entry: {
      utils: "src/utils/index.ts",
    },
    name: "utils",
    noExternal: [/find-up/],
  },
]);
