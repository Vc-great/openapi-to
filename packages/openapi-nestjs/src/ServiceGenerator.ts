import _ from "lodash";
import { Scope } from "ts-morph";

import { NestjsGenerator } from "./NestjsGenerator.ts";

import type { PluginContext } from "@openapi-to/core";
import type OasTypes from "oas/types";
import type { ClassDeclarationStructure } from "ts-morph";
import type {
  ImportDeclarationStructure,
  MethodDeclarationStructure,
  OptionalKind,
} from "ts-morph";
import type { Config } from "./types.ts";

type ImportStatementsOmitKind = Omit<ImportDeclarationStructure, "kind">;

export class ServiceGenerator extends NestjsGenerator {
  constructor(config: Config) {
    super(config);
  }

  get getResponseStatusCodes() {
    return this.openapi.response?.getResponseStatusCodes;
  }

  build(context: PluginContext): void {
    _.forEach(this.openapi.pathGroupByTag, (pathGroup, tag) => {
      const methodsStatements = _.chain(pathGroup)
        .map(({ path, method, tag }) => {
          this.operation = this.openapi.setCurrentOperation(path, method, tag);
          return this.generatorMethod();
        })
        .value();

      this.ast.createSourceFile(this.serviceFilePath, {
        statements: [
          ...this.generateImport(),
          this.generatorClass(methodsStatements),
        ],
      });
    });
  }

  generateImport(): Array<ImportDeclarationStructure> {
    const injectable: ImportStatementsOmitKind = {
      namedImports: ["Injectable"],
      moduleSpecifier: "@nestjs/common",
    };

    const repository: ImportStatementsOmitKind = {
      namedImports: [this.repositoryClassName],
      moduleSpecifier: this.repositoryModuleSpecifier,
    };

    const statements = _.chain([] as Array<ImportStatementsOmitKind>)
      .push(injectable)
      .push(repository)
      .concat(this.domainImportStatements)
      .filter(Boolean)
      .value();

    return this.ast.generateImportStatements(statements);
  }

  generatorClass(
    methodsStatements: OptionalKind<MethodDeclarationStructure>[],
  ): ClassDeclarationStructure {
    const statements = {
      name: this.serviceClassName,
      isExported: true,
      decorators: [
        {
          name: "Injectable",
        },
      ],
      docs: [],
      ctors: [
        {
          parameters: [
            {
              isReadonly: true,
              scope: Scope.Private,
              name: _.lowerFirst(this.repositoryClassName),
              type: this.repositoryClassName,
            },
          ],
        },
      ],
      methods: methodsStatements,
    };
    return this.ast.generateClassStatements(statements);
  }

  generatorMethodBody(): string {
    const pathParameters = _.chain(this.openapi.parameter?.parameters)
      .filter(["in", "path"])
      .map((item: OasTypes.ParameterObject) => _.camelCase(item.name))
      .value() as unknown as string[];

    const args = _.chain([] as string[])
      .concat(this.openapi.parameter?.hasPathParameters ? pathParameters : [])
      .concat(this.openapi.parameter?.hasQueryParameters ? "query" : [])
      .concat(this.openapi.requestBody?.hasRequestBody ? "data" : [])
      .join()
      .value();
    return `return await this.${_.lowerFirst(this.repositoryClassName)}.${this.methodName}(${args})`;
  }

  generatorMethod(): MethodDeclarationStructure {
    const statement = {
      name: this.methodName,
      decorators: [],
      parameters: this.generatorMethodParameters(),
      returnType: this.generatorReturnType(),
      docs: [],
      statements: this.generatorMethodBody(),
      isAsync: true,
    };

    return this.ast.generateMethodStatements(statement);
  }
}
