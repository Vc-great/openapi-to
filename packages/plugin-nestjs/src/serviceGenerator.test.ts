import path from "node:path";
import { dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { AST, OpenAPI } from "@openapi-to/core";

import _ from "lodash";
import Oas from "oas";
import { describe } from "vitest";

import petStore from "../mock/petstore.json";
import { ServiceGenerator } from "./ServiceGenerator.ts";

import type { OpenapiToSingleConfig } from "@openapi-to/core";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("serviceGenerator", async () => {
  const openapiToSingleConfig: OpenapiToSingleConfig = {
    name: "",
    root: "",
    input: {
      path: "",
    },
    output: {
      dir: path.resolve(__dirname, "../mock"),
    },
    plugins: [],
  };

  const context = {
    output: "",
  };

  test("service", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {};

    const controllerGenerator = new ServiceGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    controllerGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });
});
