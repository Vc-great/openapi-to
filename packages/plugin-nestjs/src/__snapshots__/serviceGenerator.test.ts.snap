// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serviceGenerator > service 1`] = `
"import { Injectable } from "@nestjs/common";
import { PetRepository } from "./repository/pet.repository";
import { FindAllQueryDto } from "./domain/findAll-query.dto";
import { FindAllUserVo } from "./domain/FindAllUser.vo";
import { testBodyDto } from "./domain/testBodyDto.dto";
import { testDto2 } from "./domain/testDto2.vo";
import { test32145 } from "./domain/test32145.vo";
import { CreateUpdateUserDto } from "./domain/CreateUpdateUserDto.dto";
import { TestidgetQueryDto } from "./domain/testidget-query.dto";
import { ApiResponse } from "./domain/ApiResponse.vo";
import { Pet } from "./domain/Pet.dto";
import { FindPetsByStatusQueryDto } from "./domain/findPetsByStatus-query.dto";
import { FindPetsByTagsQueryDto } from "./domain/findPetsByTags-query.dto";

@Injectable
export class PetService {
    constructor(private readonly petRepository: PetRepository) {
    }

    async findAll(query: FindAllQueryDto): Promise<FindAllUserVo> {
        return await this.petRepository.findAll(query)
    }

    async test(data: testBodyDto): Promise<testDto2> {
        return await this.petRepository.test(data)
    }

    async testput(data: testBodyDto): Promise<testDto2> {
        return await this.petRepository.testput(data)
    }

    async testdelete(data: number[]): Promise<test32145> {
        return await this.petRepository.testdelete(data)
    }

    async batchCreate(data: CreateUpdateUserDto[]): Promise<testDto2> {
        return await this.petRepository.batchCreate(data)
    }

    async testidget(testId: number, query: TestidgetQueryDto): Promise<testDto2> {
        return await this.petRepository.testidget(testId, query)
    }

    async uploadFile(petId: number, data: any): Promise<ApiResponse> {
        return await this.petRepository.uploadFile(petId, data)
    }

    async addPet(data: Pet): Promise<void> {
        return await this.petRepository.addPet(data)
    }

    async updatePet(data: Pet): Promise<void> {
        return await this.petRepository.updatePet(data)
    }

    async findPetsByStatus(query: FindPetsByStatusQueryDto): Promise<Pet[]> {
        return await this.petRepository.findPetsByStatus(query)
    }

    async findPetsByTags(query: FindPetsByTagsQueryDto): Promise<Pet[]> {
        return await this.petRepository.findPetsByTags(query)
    }

    async getPetById(petId: number): Promise<Pet> {
        return await this.petRepository.getPetById(petId)
    }

    async updatePetWithForm(petId: number, data: any): Promise<void> {
        return await this.petRepository.updatePetWithForm(petId, data)
    }

    async deletePet(petId: number): Promise<void> {
        return await this.petRepository.deletePet(petId)
    }
}

import { Injectable } from "@nestjs/common";
import { StoreRepository } from "./repository/store.repository";
import { Order } from "./domain/Order.dto";

@Injectable
export class StoreService {
    constructor(private readonly storeRepository: StoreRepository) {
    }

    async placeOrder(data: Order): Promise<Order> {
        return await this.storeRepository.placeOrder(data)
    }

    async getOrderById(orderId: number): Promise<Order> {
        return await this.storeRepository.getOrderById(orderId)
    }

    async deleteOrder(orderId: number): Promise<void> {
        return await this.storeRepository.deleteOrder(orderId)
    }

    async getInventory(): Promise<void> {
        return await this.storeRepository.getInventory()
    }
}

import { Injectable } from "@nestjs/common";
import { UserRepository } from "./repository/user.repository";
import { UserArray } from "./domain/UserArray.dto";
import { User } from "./domain/User.vo";
import { LoginUserQueryDto } from "./domain/loginUser-query.dto";

@Injectable
export class UserService {
    constructor(private readonly userRepository: UserRepository) {
    }

    async createUsersWithArrayInput(data: UserArray[]): Promise<void> {
        return await this.userRepository.createUsersWithArrayInput(data)
    }

    async createUsersWithListInput(data: UserArray[]): Promise<void> {
        return await this.userRepository.createUsersWithListInput(data)
    }

    async getUserByName(username: string): Promise<User> {
        return await this.userRepository.getUserByName(username)
    }

    async updateUser(username: string, data: User): Promise<void> {
        return await this.userRepository.updateUser(username, data)
    }

    async deleteUser(username: string): Promise<void> {
        return await this.userRepository.deleteUser(username)
    }

    async loginUser(query: LoginUserQueryDto): Promise<void> {
        return await this.userRepository.loginUser(query)
    }

    async logoutUser(): Promise<void> {
        return await this.userRepository.logoutUser()
    }

    async createUser(data: User): Promise<void> {
        return await this.userRepository.createUser(data)
    }
}
"
`;
